index.bs

```diff
@@ -1799,17 +1799,20 @@ Follow the <a href="https://www.w3.org/2001/tag/doc/promises-guide#one-time-even
 in the <strong><a href="https://www.w3.org/2001/tag/doc/promises-guide">Writing
 Promise-Using Specifications</a></strong> guideline.
 
-<h3 id="promises-and-events">Events should fire before Promises resolve</h3>
+<h3 id="promises-and-events">Events should fire before related Promises resolve</h3>
 
 If a Promise-based asynchronous algorithm dispatches events,
 it should dispatch them before the Promise resolves,
 rather than after.
 
-This guarantees that once the Promise resolves,
-all effects of the algorithm have been applied.
-For example, if an author changes some state
-in reaction to an event which the Promise dispatches,
-they can be sure that all of the state is consistent if the Promise is resolved.
+When a promise is resolved, a microtask is queued to run its reaction callbacks.
```

Comment:
I think so.  Though when you say "all event-related callbacks" is that really the case?  Because the text then says that the second callback goes after the promise, because the promise reactions run when the stack drains.

(I guess I'm still not 100% on all of this.  A reference might still be necessary.)

Reply:
> Though when you say "all event-related callbacks" is that really the case?

Yes.

> Because the text then says that the second callback goes after the promise, because the promise reactions run when the stack drains.

Right, but that doesn't prevent running "all event-related callbacks" as part of the dispatch, it just additionally involves the promise reaction callbacks.

Dispatching the event involves calling callbacks, which means the JS stack populates and empties, which means microtasks run, which means promise callbacks run 

> (I guess I'm still not 100% on all of this. A reference might still be necessary.)

By reference, do you mean a link to another spec? Would any of the links I've provided so far work as this reference?

Reply:
OK, so I'm following you now, but none of this detail is necessarily obvious.  Citing one of those documents might help fill in some details if people (like me) find themselves struggling.

Dispatching an event is synchronous.  All event callbacks are invoked in turn.  In comparison, resolving a promise creates a microtask for each callback.  Microtasks do not run immediately, but run when the stack drains.  Mixing these patterns creates a potentially surprising interaction.  Invoking any callback means populating and then draining the stack.  When the stack drains, any microtasks that are dependent on promises are invoked.  Thus, if an event is dispatched after a promise resolves, the first event callback will be invoked.  However, when the callback returns, the stack drains and then callbacks for the promise run as microtasks.  Those microtasks run before the second and subsequent event callbacks.

Reply:
> resolving a promise creates a microtask for each callback.

It creates a microtask to run all currently queued callbacks. All of the callbacks will run in the same microtask. Although, I'm not 100% sure that difference is observable.

> When the stack drains, any microtasks that are dependent on promises are invoked

Yes, but _all_ microtasks run, not just promise-related ones. Microtasks queued for, say, mutation observers, will also run.

I'll link up the definitions for dispatch and resolve when I get time (I don't have a bikeshed environment on my new machine, so I'll need to set that up)